### ISSUE TITLE
Matrix.rank() incorrect results

### ISSUE THREAD

**Issue by siefkenj** [sympy/sympy#15872]
Another example where a matrix rank is computed incorrectly:
As referenced in issue #15858 the following matrix rank is not computed correctly:

```
In [20]: A = Matrix([[1, 1, 1, 0], [-2, -1, 0, -1], [0, 0, -1, -1], [0, 0, 2, 1]
    ...: ])

In [21]: B = A - Matrix.eye(4)*I

In [22]: B.rank()
Out[22]: 3

In [23]: (B**2).rank()
Out[23]: 2

In [24]: (B**3).rank()
Out[24]: 3

In [25]: (B**3).rank(simplify=True)
Out[25]: 3

In [26]: simplify(B**3).rank()
Out[26]: 2
```

The rank of `B**3` should be 2. Further, passing the `simplify` flag doesn't help, but applying `simplify()` to the whole expression does. This might have something to do with issue #13298 causing row reduction to make incorrect assumptions.

**Comment by ethankward** [comment]
Tracing back the issue, one thing that jumps out is

```
>>> e = -128*I + 64*(1 + I)**2
>>> e.is_zero
False
```

**Comment by sylee957** [comment]
SymPy does not expand multiplications automatically, and I think it's making things complicated.

![image](https://user-images.githubusercontent.com/34944973/51873412-68b5df80-23a0-11e9-8b12-a11f36a7344e.png)

I think real numbers are not treated unevaluated like 1 * 2 * 3, but imaginary units are treated like polynomials.
But I think those things would be too philosophical to discuss here, because other CAS may actually use different approach of defining complex expression, than SymPy's way of treating it like a polynomial with `S.ImaginaryUnit`
And also it can be relevant to the reason that numbers are not being absorbed by floating point expressions in #15844.

One thing I have found from `fft()` in discrete module, was that it had been calling `expand` every time when multiplication occurs to simplify the complex expression, though for generic matrix multiplication other than such linear transformations, it may only be guaranteed to bring more stability for pure numeric cases without free symbols.

Though there may need some work to be done inside `rref`, as I suppose that there may be some issue for `_find_reasonable_pivot` not granting much better numeric stability than applying some priori simplification, than the complexity of code it brings

**Comment by sylee957** [comment]
I think this issue had been fixed after merging #15875

**Comment by oscarbenjamin** [comment]
This example also came up in #17837 which I came to while investigating https://github.com/sympy/sympy/issues/14675#issuecomment-548957534
